src/frontend/screens/Settings/ProjectSettings/DeviceName.tsx

```diff
@@ -0,0 +1,144 @@
+    defaultMessage: 'Device Name',
+  },
+  yourDevice: {
+    id: 'Screens.Settings.ProjectSettings.DeviceName.yourDevice',
+    defaultMessage: 'Your Device Name',
+  },
+  editDevice: {
+    id: 'Screens.Settings.ProjectSettings.DeviceName.editDevice',
+    defaultMessage: 'Edit Device Name',
+  },
+  discardChanges: {
+    id: 'Screens.Settings.ProjectSettings.DeviceName.discardChanges',
+    defaultMessage: 'Discard Changes',
+  },
+  continueEditting: {
+    id: 'Screens.Settings.ProjectSettings.DeviceName.continueEditting',
+    defaultMessage: 'Continue Editting',
+  },
+});
+
+type FormData = {
+  name: string;
+};
+
+export const DeviceName: NativeNavigationComponent<'DeviceName'> = ({
+  navigation,
+}) => {
+  const [isEditting, setIsEditting] = React.useState(false);
+  const deviceInfo = useDeviceInfo();
+  const {mutate, isPending, variables: newDeviceName} = useEditDeviceInfo();
+  const {formatMessage: t} = useIntl();
+  const {
+    control,
+    handleSubmit,
+    watch,
+    formState: {errors},
+  } = useForm<FormData>({defaultValues: {name: ''}});
+
+  React.useEffect(() => {
+    navigation.setOptions({
+      headerRight: () => (
+        <HeaderRightDeviceName
+          handleSubmit={handleSubmit(val => {
+            mutate(val.name);
+            setIsEditting(false);
+          })}
+          setIsEditting={setIsEditting}
+          isEditting={isEditting}
+        />
+      ),
+    });
+  }, [navigation, isEditting, handleSubmit, setIsEditting]);
+
+  React.useEffect(() => {
+    const unsubscribe = navigation.addListener('beforeRemove', e => {
+      if (!isEditting) {
+        // if user is not doing anything then don't intercept back action
+        return;
+      }
+
+      // Prevent default behavior of leaving the screen
+      e.preventDefault();
+
+      if (watch().name.length < 1) {
+        setIsEditting(false);
+        return;
+      }
+      // Prompt the user before leaving the screen
+      Alert.alert(t(m.discardChanges), undefined, [
+        {text: t(m.continueEditting), style: 'cancel', onPress: () => {}},
+        {
+          text: t(m.discardChanges),
+          onPress: () => navigation.dispatch(e.data.action),
+        },
+      ]);
+    });
+
+    return () => unsubscribe();
+  }, [navigation, isEditting, watch]);
+
+  return (
+    <TouchableWithoutFeedback
+      style={styles.container}
+      onPress={() => Keyboard.dismiss()}>
+      <Text style={{fontWeight: 'bold', marginBottom: 10}}>
+        {t(!isEditting ? m.yourDevice : m.editDevice)}
+      </Text>
+      {!isEditting ? (
+        deviceInfo.isLoading ? (
+          <Loading />
+        ) : (
+          // Optimistically updates the device name
+          <Text>{isPending ? newDeviceName : deviceInfo.data?.name}</Text>
+        )
+      ) : (
+        <React.Fragment>
+          <HookFormTextInput
+            name="name"
+            control={control}
+            rules={{required: true, maxLength: 60}}
```

Comment:
How are validation errors displayed to the user? E.g. what would happen if they entered a name over 60 characters? And what would happen if they tried to delete the device name and save?

Reply:
[We highlight the input box as red](https://github.com/digidem/CoMapeo-mobile/blob/a7bfb675c0585cc81eaca3872dc334e6519242fc/src/frontend/screens/Settings/ProjectSettings/DeviceName.tsx#L123) if there is an error